File: backend/scripts_iniciales/setup-db.js
const { MongoClient } = require('mongodb');

const MONGODB_URI = 'mongodb://mongo:TU_PASSWORD@ballast.proxy.rlwy.net:12247/geocrimen_tacna';

async function setupDB() {
  const client = new MongoClient(MONGODB_URI);

  try {
    await client.connect();
    console.log('✅ Conectado a MongoDB Railway');

    const db = client.db('geocrimen_tacna');

    // ── 1. USUARIOS ────────────────────────────────────────────────────────
    await db.createCollection('usuarios', {
      validator: {
        $jsonSchema: {
          bsonType: 'object',
          required: ['nombre', 'email', 'password_hash', 'rol'],
          properties: {
            nombre:        { bsonType: 'string' },
            email:         { bsonType: 'string' },
            password_hash: { bsonType: 'string' },
            rol:           { enum: ['ciudadano', 'policia', 'admin'] },
            telefono:      { bsonType: 'string' },
            ubicacion_default: {
              bsonType: 'object',
              properties: {
                type:        { enum: ['Point'] },
                coordinates: { bsonType: 'array' },
              },
            },
            distrito:  { bsonType: 'string' },
            ubigeo:    { bsonType: 'string' },
            activo:    { bsonType: 'bool' },
            creado_en: { bsonType: 'date' },
          },
        },
      },
    });
    await db.collection('usuarios').createIndexes([
      { key: { email: 1 },              name: 'email_unique',    unique: true },
      { key: { ubicacion_default: '2dsphere' }, name: 'ubicacion_2dsphere' },
      { key: { ubigeo: 1 },             name: 'ubigeo' },
      { key: { rol: 1 },                name: 'rol' },
    ]);
    console.log('✅ Colección usuarios creada');

    // ── 2. REPORTES_CIUDADANO ──────────────────────────────────────────────
    await db.createCollection('reportes_ciudadano', {
      validator: {
        $jsonSchema: {
          bsonType: 'object',
          required: ['sub_tipo', 'modalidad', 'ubicacion', 'distrito', 'fecha_hecho', 'estado'],
          properties: {
            usuario_id: { bsonType: ['objectId', 'null'] },
            anonimo:    { bsonType: 'bool' },
            tipo:       { enum: ['PATRIMONIO (DELITO)'] },
            sub_tipo:   { enum: ['HURTO', 'ROBO', 'EXTORSION'] },
            modalidad:  { bsonType: 'string' },
            ubicacion: {
              bsonType: 'object',
              required: ['type', 'coordinates'],
              properties: {
                type:        { enum: ['Point'] },
                coordinates: { bsonType: 'array' },
              },
            },
            direccion:    { bsonType: 'string' },
            distrito:     { bsonType: 'string' },
            ubigeo:       { bsonType: 'string' },
            fecha_hecho:  { bsonType: 'date' },
            hora_aprox:   { bsonType: 'string' },
            descripcion:  { bsonType: 'string' },
            fotos:        { bsonType: 'array' },
            estado:       { enum: ['pendiente', 'confirmado', 'rechazado'] },
            revisado_por: { bsonType: ['objectId', 'null'] },
            nota_revision:{ bsonType: 'string' },
            incidente_id: { bsonType: ['objectId', 'null'] },
            creado_en:    { bsonType: 'date' },
          },
        },
      },
    });
    await db.collection('reportes_ciudadano').createIndexes([
      { key: { estado: 1 },      name: 'estado' },
      { key: { ubigeo: 1 },      name: 'ubigeo' },
      { key: { creado_en: -1 },  name: 'creado_en_desc' },
    ]);
    console.log('✅ Colección reportes_ciudadano creada');

    // ── 3. INCIDENTES ──────────────────────────────────────────────────────
    await db.createCollection('incidentes', {
      validator: {
        $jsonSchema: {
          bsonType: 'object',
          required: ['fuente', 'sub_tipo', 'modalidad', 'ubicacion', 'distrito', 'fecha_hecho', 'anio', 'mes'],
          properties: {
            fuente:       { enum: ['ciudadano', 'policia', 'sidpol'] },
            reporte_id:   { bsonType: ['objectId', 'null'] },
            verificado_por: { bsonType: ['objectId', 'null'] },
            tipo:         { enum: ['PATRIMONIO (DELITO)'] },
            sub_tipo:     { enum: ['HURTO', 'ROBO', 'EXTORSION'] },
            modalidad:    { bsonType: 'string' },
            ubicacion: {
              bsonType: 'object',
              required: ['type', 'coordinates'],
              properties: {
                type:        { enum: ['Point'] },
                coordinates: { bsonType: 'array' },
              },
            },
            direccion:   { bsonType: 'string' },
            distrito:    { bsonType: 'string' },
            ubigeo:      { bsonType: 'string' },
            fecha_hecho: { bsonType: 'date' },
            anio:        { bsonType: 'int' },
            mes:         { bsonType: 'int', minimum: 1, maximum: 12 },
            descripcion: { bsonType: 'string' },
            fotos:       { bsonType: 'array' },
            creado_en:   { bsonType: 'date' },
          },
        },
      },
    });
    await db.collection('incidentes').createIndexes([
      { key: { ubicacion: '2dsphere' },        name: 'ubicacion_2dsphere' },
      { key: { ubigeo: 1, anio: 1, mes: 1 },   name: 'ubigeo_anio_mes' },
      { key: { sub_tipo: 1 },                  name: 'sub_tipo' },
      { key: { fuente: 1 },                    name: 'fuente' },
      { key: { fecha_hecho: -1 },              name: 'fecha_hecho_desc' },
    ]);
    console.log('✅ Colección incidentes creada');

    // ── 4. ESTADISTICAS_SIDPOL ─────────────────────────────────────────────
    await db.createCollection('estadisticas_sidpol', {
      validator: {
        $jsonSchema: {
          bsonType: 'object',
          required: ['anio', 'mes', 'ubigeo', 'distrito', 'cantidad'],
          properties: {
            anio:         { bsonType: 'int' },
            mes:          { bsonType: 'int', minimum: 1, maximum: 12 },
            ubigeo:       { bsonType: 'string' },
            departamento: { bsonType: 'string' },
            provincia:    { bsonType: 'string' },
            distrito:     { bsonType: 'string' },
            tipo:         { bsonType: 'string' },
            sub_tipo:     { bsonType: 'string' },
            modalidad:    { bsonType: 'string' },
            cantidad:     { bsonType: 'int' },
            importado_en: { bsonType: 'date' },
          },
        },
      },
    });
    await db.collection('estadisticas_sidpol').createIndexes([
      {
        key: { ubigeo: 1, anio: 1, mes: 1, sub_tipo: 1, modalidad: 1 },
        name: 'unique_sidpol',
        unique: true,
      },
      { key: { ubigeo: 1 }, name: 'ubigeo' },
    ]);
    console.log('✅ Colección estadisticas_sidpol creada');

    // ── 5. ZONAS_RIESGO ────────────────────────────────────────────────────
    await db.createCollection('zonas_riesgo', {
      validator: {
        $jsonSchema: {
          bsonType: 'object',
          required: ['centroide', 'radio_metros', 'distrito', 'nivel_riesgo', 'total_incidentes', 'tendencia'],
          properties: {
            centroide: {
              bsonType: 'object',
              required: ['type', 'coordinates'],
              properties: {
                type:        { enum: ['Point'] },
                coordinates: { bsonType: 'array' },
              },
            },
            radio_metros:         { bsonType: 'int' },
            distrito:             { bsonType: 'string' },
            ubigeo:               { bsonType: 'string' },
            nivel_riesgo:         { enum: ['bajo', 'medio', 'alto', 'critico'] },
            total_incidentes:     { bsonType: 'int' },
            delito_predominante:  { bsonType: 'string' },
            tendencia:            { enum: ['subiendo', 'estable', 'bajando'] },
            periodo_analizado: {
              bsonType: 'object',
              properties: {
                desde: { bsonType: 'date' },
                hasta: { bsonType: 'date' },
              },
            },
            calculado_en: { bsonType: 'date' },
          },
        },
      },
    });
    await db.collection('zonas_riesgo').createIndexes([
      { key: { centroide: '2dsphere' }, name: 'centroide_2dsphere' },
      { key: { nivel_riesgo: 1 },       name: 'nivel_riesgo' },
      { key: { ubigeo: 1 },             name: 'ubigeo' },
    ]);
    console.log('✅ Colección zonas_riesgo creada');

    // ── 6. ALERTAS ─────────────────────────────────────────────────────────
    await db.createCollection('alertas', {
      validator: {
        $jsonSchema: {
          bsonType: 'object',
          required: ['usuario_id', 'tipo', 'mensaje'],
          properties: {
            usuario_id:  { bsonType: 'objectId' },
            incidente_id:{ bsonType: ['objectId', 'null'] },
            zona_id:     { bsonType: ['objectId', 'null'] },
            tipo:        { enum: ['nuevo_incidente', 'zona_peligrosa', 'zona_actualizada'] },
            mensaje:     { bsonType: 'string' },
            leida:       { bsonType: 'bool' },
            push_enviado:{ bsonType: 'bool' },
            creado_en:   { bsonType: 'date' },
          },
        },
      },
    });
    await db.collection('alertas').createIndexes([
      { key: { usuario_id: 1, leida: 1 }, name: 'usuario_leida' },
      { key: { creado_en: -1 },           name: 'creado_en_desc' },
    ]);
    console.log('✅ Colección alertas creada');

    // ── RESUMEN ────────────────────────────────────────────────────────────
    const colecciones = await db.listCollections().toArray();
    console.log(`\n🎉 Base de datos lista. Colecciones creadas: ${colecciones.length}`);
    colecciones.forEach(c => console.log(`   • ${c.name}`));

  } catch (err) {
    if (err.codeName === 'NamespaceExists') {
      console.log(`⚠️  La colección ya existe, omitiendo: ${err.message}`);
    } else {
      console.error('❌ Error:', err.message);
    }
  } finally {
    await client.close();
    console.log('\n🔌 Desconectado.');
  }
}

setupDB();

File: backend/scripts_iniciales/setup_db.py
import os
import sys
from datetime import datetime
from pymongo import MongoClient
from dotenv import load_dotenv

# Permitir importaciones desde el directorio principal
sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
from utils.crypto import hash_password

# Cargar configuración desde .env
load_dotenv()

MONGO_URL = os.getenv("MONGO_URL")

def setup_db():
    if not MONGO_URL:
        print("❌ Error: Faltan las credenciales MONGO_URL en el archivo .env")
        return

    client = MongoClient(MONGO_URL)
    db = client['geocrimen_tacna']

    print("✅ Conectado a MongoDB Railway")

    # 1. USUARIOS
    try:
        db.create_collection('usuarios', validator={
            "$jsonSchema": {
                "bsonType": "object",
                "required": ["nombre", "email", "password_hash", "rol"],
                "properties": {
                    "nombre": {"bsonType": "string"},
                    "email": {"bsonType": "string"},
                    "password_hash": {"bsonType": "string"},
                    "rol": {"enum": ["ciudadano", "policia", "admin"]},
                    "telefono": {"bsonType": "string"},
                    "ubicacion_default": {
                        "bsonType": "object",
                        "properties": {
                            "type": {"enum": ["Point"]},
                            "coordinates": {"bsonType": "array"}
                        }
                    },
                    "distrito": {"bsonType": "string"},
                    "ubigeo": {"bsonType": "string"},
                    "activo": {"bsonType": "bool"},
                    "creado_en": {"bsonType": "date"}
                }
            }
        })
        db.usuarios.create_index("email", unique=True)
        db.usuarios.create_index([("ubicacion_default", "2dsphere")])
        print("✅ Colección 'usuarios' creada")
    except Exception as e:
        print("⚠️  La colección 'usuarios' ya existe o hubo un error:", e)

    # 2. REPORTES_CIUDADANO
    try:
        db.create_collection('reportes_ciudadano', validator={
            "$jsonSchema": {
                "bsonType": "object",
                "required": ["tipo_hecho", "subtipo_hecho", "ubicacion", "departamento_hecho", "provincia_hecho", "distrito_hecho", "estado"],
                "properties": {
                    "usuario_id": {"bsonType": ["objectId", "null"]},
                    "anonimo": {"bsonType": "bool"},
                    "tipo_hecho": {"enum": ["PATRIMONIO (DELITO)"]},
                    "subtipo_hecho": {"enum": ["HURTO", "ROBO"]},
                    "modalidad_hecho": {"bsonType": "string"},
                    "ubicacion": {
                        "bsonType": "object",
                        "required": ["type", "coordinates"],
                        "properties": {
                            "type": {"enum": ["Point"]},
                            "coordinates": {"bsonType": "array"}
                        }
                    },
                    "direccion_hecho": {"bsonType": "string"},
                    "departamento_hecho": {"bsonType": "string"},
                    "provincia_hecho": {"bsonType": "string"},
                    "distrito_hecho": {"bsonType": "string"},
                    "fecha_hora_hecho": {"bsonType": "date"},
                    "descripcion": {"bsonType": "string"},
                    "fotos": {"bsonType": "array"},
                    "estado": {"enum": ["pendiente", "confirmado", "rechazado"]},
                    "creado_en": {"bsonType": "date"}
                }
            }
        })
        print("✅ Colección 'reportes_ciudadano' creada")
    except Exception as e:
        print("⚠️  La colección 'reportes_ciudadano' ya existe o hubo un error.")

    # 3. HISTORIAL_DELITOS 
    # Almacena de forma estandarizada los delitos (ArcGIS y Reportes Ciudadanos validados)
    try:
        db.create_collection('historial_delitos', validator={
            "$jsonSchema": {
                "bsonType": "object",
                "required": ["ubicacion", "departamento_hecho", "provincia_hecho", "distrito_hecho", "fecha_hecho", "tipo_hecho", "subtipo_hecho"],
                "properties": {
                    "ubicacion": {
                        "bsonType": "object",
                        "required": ["type", "coordinates"],
                        "properties": {
                            "type": {"enum": ["Point"]},
                            "coordinates": {"bsonType": "array"}
                        }
                    },
                    "direccion_hecho": {"bsonType": ["string", "null"]},
                    "tipo_via": {"bsonType": ["string", "null"]},
                    "departamento_hecho": {"bsonType": ["string", "null"]},
                    "provincia_hecho": {"bsonType": ["string", "null"]},
                    "distrito_hecho": {"bsonType": ["string", "null"]},
                    "ubigeo": {"bsonType": ["string", "null"]},
                    "fecha_hecho": {"bsonType": "date"},
                    "turno_hecho": {"bsonType": ["string", "null"]},
                    "tipo_hecho": {"bsonType": ["string", "null"]},
                    "subtipo_hecho": {"bsonType": ["string", "null"]},
                    "modalidad_hecho": {"bsonType": ["string", "null"]},
                    "estado_coord": {"bsonType": ["string", "null"]}, 
                    "fuente": {"enum": ["arcgis_sidpol", "ciudadano"]},
                    "creado_en": {"bsonType": "date"}
                }
            }
        })
        db.historial_delitos.create_index([("ubicacion", "2dsphere")])
        print("✅ Colección 'historial_delitos' creada")
    except Exception as e:
        print("⚠️  La colección 'historial_delitos' ya existe o hubo un error.")

    # 4. ZONAS_RIESGO
    try:
        db.create_collection('zonas_riesgo', validator={
            "$jsonSchema": {
                "bsonType": "object",
                "required": ["centroide", "radio_metros", "distrito", "nivel_riesgo", "total_incidentes", "tendencia"],
                "properties": {
                    "centroide": {
                        "bsonType": "object",
                        "required": ["type", "coordinates"],
                        "properties": {
                            "type": {"enum": ["Point"]},
                            "coordinates": {"bsonType": "array"}
                        }
                    },
                    "radio_metros": {"bsonType": "int"},
                    "distrito": {"bsonType": "string"},
                    "ubigeo": {"bsonType": "string"},
                    "nivel_riesgo": {"enum": ["bajo", "medio", "alto", "critico"]},
                    "total_incidentes": {"bsonType": "int"},
                    "delito_predominante": {"bsonType": "string"},
                    "tendencia": {"enum": ["subiendo", "estable", "bajando"]},
                    "periodo_analizado": {
                        "bsonType": "object",
                        "properties": {
                            "desde": {"bsonType": "date"},
                            "hasta": {"bsonType": "date"}
                        }
                    },
                    "calculado_en": {"bsonType": "date"}
                }
            }
        })
        db.zonas_riesgo.create_index([("centroide", "2dsphere")])
        print("✅ Colección 'zonas_riesgo' creada")
    except Exception as e:
        print("⚠️  La colección 'zonas_riesgo' ya existe o hubo un error.")

    # Crear Usuarios Iniciales
    print("\n--- CREANDO USUARIOS DE PRUEBA Y ADMIN ---")
    
    import datetime as dt
    now_utc = dt.datetime.now(dt.timezone.utc)
    
    usuarios_iniciales = [
        {
            "nombre": "Admin Supremo",
            "email": "admin@sgeo.com",
            "password_hash": hash_password("ffe.Ad95"),
            "rol": "admin",
            "activo": True,
            "creado_en": now_utc
        },
        {
            "nombre": "Agente Policia 1",
            "email": "policia@sgeo.com",
            "password_hash": hash_password("123456"),
            "rol": "policia",
            "activo": True,
            "creado_en": now_utc
        },
        {
            "nombre": "Ciudadano Juan",
            "email": "juan@ciudadano.com",
            "password_hash": hash_password("123456"),
            "rol": "ciudadano",
            "activo": True,
            "creado_en": now_utc
        }
    ]

    for user in usuarios_iniciales:
        # Verifica si ya existe para no duplicarlo cada vez
        if not db.usuarios.find_one({"email": user["email"]}):
            db.usuarios.insert_one(user)
            print(f"👤 Creado: {user['email']} ({user['rol']})")
        else:
            print(f"🔹 Ya existía: {user['email']}")

    print("\n🎉 Base de datos lista.")

if __name__ == "__main__":
    setup_db()